 # **Shopify Winter 2022 Data Science Intern Challenge**

**Question 1:** Given some sample data, write a program to answer the following: click here to access the required data set

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis.

**a. Think about what could be going wrong with our calculation. Think about a better way to evaluate this data.**

In [2]:
import numpy as np 
import pandas as pd 
import seaborn as sns
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))


/kaggle/input/shopify-data-science-internship-challenge/Shopify.csv


In [3]:
data = pd.read_csv("../input/shopify-data-science-internship-challenge/Shopify.csv")

In [11]:
data.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [56]:
data.isnull()

order_id  shop_id  user_id  order_amount  total_items  payment_method  \
0        False    False    False         False        False           False   
1        False    False    False         False        False           False   
2        False    False    False         False        False           False   
3        False    False    False         False        False           False   
4        False    False    False         False        False           False   
...        ...      ...      ...           ...          ...             ...   
4995     False    False    False         False        False           False   
4996     False    False    False         False        False           False   
4997     False    False    False         False        False           False   
4998     False    False    False         False        False           False   
4999     False    False    False         False        False           False   

      created_at  item_price  
0          False       False  
1          False       False  
2          False       False  
3          False       False  
4          False       False  
...          ...         ...  
4995       False       False  
4996       False       False  
4997       False       False  
4998       False       False  
4999       False       False  

[5000 rows x 8 columns]

In [61]:
data.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4995    False
4996    False
4997    False
4998    False
4999    False
Length: 5000, dtype: bool

It is evident that, there are **no null values** and also **no repeated/duplicated values**.

In [4]:
data.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

From this, we can cross check that, there are 5000 order_ids, 100 shop_ids, user_ids from 607 to 999, minimum order_amount of 90 to maximum order_amount of 70400.

We can also see here that there is a standard deviation 41282.5 in order amount, which is considered as an outlier of this data and maximum order amount of 704000 is another outlier in this data, which made mean to be 3145.13.

In [13]:
data['item_price'] = data['order_amount']/data['total_items']
data.describe()

order_id      shop_id      user_id   order_amount  total_items  \
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000   
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720   
std    1443.520003    29.006118    87.798982   41282.539349    116.32032   
min       1.000000     1.000000   607.000000      90.000000      1.00000   
25%    1250.750000    24.000000   775.000000     163.000000      1.00000   
50%    2500.500000    50.000000   849.000000     284.000000      2.00000   
75%    3750.250000    75.000000   925.000000     390.000000      3.00000   
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000   

         item_price  
count   5000.000000  
mean     387.742800  
std     2441.963725  
min       90.000000  
25%      133.000000  
50%      153.000000  
75%      169.000000  
max    25725.000000

Here it is evident that some items are sold at a price tag of **25725 per piece** and **order amount of 704000**. Which are huge!! 

In [70]:
huge_order = data[data['order_amount'] == 704000]
huge_order

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  item_price  
15    2017-03-07 4:00:00       352.0  
60    2017-03-04 4:00:00       352.0  
520   2017-03-02 4:00:00       352.0  
1104  2017-03-24 4:00:00       352.0  
1362  2017-03-15 4:00:00       352.0  
1436  2017-03-11 4:00:00       352.0  
1562  2017-03-19 4:00:00       352.0  
1602  2017-03-17 4:00:00       352.0  
2153  2017-03-12 4:00:00       352.0  
2297  2017-03-07 4:00:00       352.0  
2835  2017-03-28 4:00:00       352.0  
2969  2017-03-28 4:00:00       352.0  
3332  2017-03-24 4:00:00       352.0  
4056  2017-03-28 4:00:00       352.0  
4646  2017-03-02 4:00:00       352.0  
4868  2017-03-22 4:00:00       352.0  
4882  2017-03-25 4:00:00       352.0

All these huge orders are from the **shop_id = 42** and **user_id = 607**. We can suspect this as a fraud.

In [71]:
overpriced_orders = data[data['item_price'] == 25725]
overpriced_orders

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1056      1057       78      800         25725            1          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1259      1260       78      775         77175            3    credit_card   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1452      1453       78      812         25725            1    credit_card   
1529      1530       78      810         51450            2           cash   
2270      2271       78      855         25725            1    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2773      2774       78      890         25725            1           cash   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2906      2907       78      817         77175            3          debit   
2922      2923       78      740         25725            1          debit   
3085      3086       78      910         25725            1           cash   
3101      3102       78      855         51450            2    credit_card   
3151      3152       78      745         25725            1    credit_card   
3167      3168       78      927         51450            2           cash   
3403      3404       78      928         77175            3          debit   
3440      3441       78      982         25725            1          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
3780      3781       78      889         25725            1           cash   
4040      4041       78      852         25725            1           cash   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4505      4506       78      866         25725            1          debit   
4584      4585       78      997         25725            1           cash   
4715      4716       78      818         77175            3          debit   
4918      4919       78      823         25725            1           cash   

               created_at  item_price  
160    2017-03-12 5:56:57     25725.0  
490   2017-03-26 17:08:19     25725.0  
493   2017-03-16 21:39:35     25725.0  
511    2017-03-09 7:23:14     25725.0  
617   2017-03-18 11:18:42     25725.0  
691   2017-03-27 22:51:43     25725.0  
1056  2017-03-15 10:16:45     25725.0  
1193  2017-03-16 16

In [22]:
overpriced_orders.total_items.count()

46

There are **46 orders with each product costs as much as 25725** and all are from **shop_id = 78**. This is the place where the scam happened.

In [68]:
final_data = data[data['shop_id'] != 78]
final_data.describe()

order_id      shop_id      user_id   order_amount  total_items  \
count  4954.000000  4954.000000  4954.000000    4954.000000  4954.000000   
mean   2498.990916    49.819540   848.919257    2717.367784     8.851029   
std    1444.498907    29.014845    87.846007   41155.996469   116.857286   
min       1.000000     1.000000   607.000000      90.000000     1.000000   
25%    1248.250000    24.000000   775.000000     163.000000     1.000000   
50%    2494.500000    50.000000   849.000000     284.000000     2.000000   
75%    3750.750000    74.000000   925.000000     390.000000     3.000000   
max    5000.000000   100.000000   999.000000  704000.000000  2000.000000   

        item_price  
count  4954.000000  
mean    152.475575  
std      31.260218  
min      90.000000  
25%     132.000000  
50%     153.000000  
75%     168.000000  
max     352.000000

By removing the **shop_id = 78**, we got the final_data with **maximum price of the product as 352**. Which is a good sign for us according to the peoblem. But still maximum total items shows 2000.

In [69]:
new_data = final_data[final_data['user_id'] != 607]
new_data.describe()

order_id      shop_id      user_id  order_amount  total_items  \
count  4937.000000  4937.000000  4937.000000   4937.000000  4937.000000   
mean   2499.551347    49.846465   849.752279    302.580514     1.994734   
std    1444.069407    29.061131    86.840313    160.804912     0.982821   
min       1.000000     1.000000   700.000000     90.000000     1.000000   
25%    1248.000000    24.000000   775.000000    163.000000     1.000000   
50%    2497.000000    50.000000   850.000000    284.000000     2.000000   
75%    3751.000000    74.000000   925.000000    387.000000     3.000000   
max    5000.000000   100.000000   999.000000   1760.000000     8.000000   

        item_price  
count  4937.000000  
mean    151.788536  
std      29.034215  
min      90.000000  
25%     132.000000  
50%     153.000000  
75%     166.000000  
max     352.000000

By removing the **user_id = 607**,  we got the new_data with **maximum total items as 8** and a **mean order amount of 302.6** with a **standard deviation of 160.8** and a **median of 284**. Which are far better than the initial AOV of 3145.13 and 41282.5.

**b. What metric would you report for this dataset?**

It is better to use the data without any outliers and with an **average order value** after removing the shop_id 78 and user_id 607.

**c. What is its value?**

**AOV = 302.6 dollars**

**Question 2:** For this question you’ll need to use SQL. Follow this link to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below.

**a. How many orders were shipped by Speedy Express in total?**

In [ ]:
SELECT COUNT(OrderID) 
FROM Orders
WHERE ShipperID = (SELECT ShipperID 
                   FROM Shippers
                   WHERE ShipperName = 'Speedy Express');

**54**

**b. What is the last name of the employee with the most orders?**

In [ ]:
SELECT Employees.LastName, COUNT(*) MostOrders
FROM Orders

JOIN Employees
ON Orders.EmployeeID = Employees.EmployeeID

GROUP BY Employees.LastName
ORDER BY MostOrders DESC
LIMIT 1

**Peacock	40**

**c. What product was ordered the most by customers in Germany?**

In [ ]:
SELECT prod.ProductName, SUM(orddet.Quantity) MostOrdered
FROM OrderDetails orddet

JOIN Orders ord 
ON ord.OrderID = orddet.OrderID

JOIN Customers cust 
ON cust.CustomerID = ord.CustomerID

JOIN Products prod 
ON prod.ProductID = orddet.ProductID

WHERE cust.Country = 'Germany'
GROUP BY ProductName
ORDER BY MostOrdered DESC

**Boston Crab Meat	160**